app/master/cluster_master.py

```diff
@@ -105,22 +105,39 @@ def get_slave(self, slave_id=None, slave_url=None):
 
         raise ItemNotFoundError('Requested slave ({}) does not exist.'.format(slave_id))
 
-    def connect_new_slave(self, slave_url, num_executors):
+    def connect_slave(self, slave_url, num_executors):
         """
-        Add a new slave to this master.
+        Connect a slave to this master.
+
+        If this slave has already been registered with the master and isn't working on a build, make sure that slave
+        is marked as alive and ready to accept builds.
 
         :type slave_url: str
         :type num_executors: int
-        :return: The slave id of the new slave
-        :rtype: int
-        """
-        slave = Slave(slave_url, num_executors)
-        self._all_slaves_by_url[slave_url] = slave
-        self._slave_allocator.add_idle_slave(slave)
-
-        self._logger.info('Slave on {} connected to master with {} executors. (id: {})',
-                          slave_url, num_executors, slave.id)
-        return {'slave_id': str(slave.id)}
+        :return: The response with the slave id of the slave.
+        :rtype: dict[str, str]
+        """
+        if slave_url in self._all_slaves_by_url:
+            self._logger.warning('Slave on {} requested to connect to master, even though already connected', slave_url)
+            slave = self._all_slaves_by_url.get(slave_url)
+            slave.set_is_alive(True)
+            if slave.current_build_id is None:
```

Comment:
If the slave is hard killed while working on a build (so `slave.current_build_id` is not None), and then reconnected relatively soon (before the master detects that the slave died), then it seems like the slave might be stuck after that (and whatever build it was working on might be stuck to) since from the master's perspective the slave is still working on that build?